File: core/services/schedule_service.py
from datetime import datetime, timedelta
from django.utils import timezone
from core.models import Area, AreaSchedule


class ScheduleService:
    """Servicio para manejar horarios de áreas y empleados"""
    
    @staticmethod
    def get_expected_times(area, date):
        """
        Obtiene las horas esperadas de entrada y salida para un área en una fecha específica
        
        Args:
            area: Instancia del modelo Area
            date: Fecha para la cual obtener el horario
            
        Returns:
            tuple: (hora_entrada_esperada, hora_salida_esperada) o (None, None) si no hay horario
        """
        if not hasattr(area, 'schedule'):
            return None, None
        
        schedule = area.schedule
        weekday = date.weekday()  # 0=Lunes, 6=Domingo
        
        if weekday == 0 and schedule.monday_active:
            return schedule.monday_start, schedule.monday_end
        elif weekday == 1 and schedule.tuesday_active:
            return schedule.tuesday_start, schedule.tuesday_end
        elif weekday == 2 and schedule.wednesday_active:
            return schedule.wednesday_start, schedule.wednesday_end
        elif weekday == 3 and schedule.thursday_active:
            return schedule.thursday_start, schedule.thursday_end
        elif weekday == 4 and schedule.friday_active:
            return schedule.friday_start, schedule.friday_end
        elif weekday == 5 and schedule.saturday_active:
            return schedule.saturday_start, schedule.saturday_end
        elif weekday == 6 and schedule.sunday_active:
            return schedule.sunday_start, schedule.sunday_end
        
        return None, None
    
    @staticmethod
    def create_default_schedule(area):
        """
        Crea un horario por defecto para un área (8:00 AM - 5:00 PM, L-V)
        
        Args:
            area: Instancia del modelo Area
            
        Returns:
            AreaSchedule: Instancia del horario creado
        """
        from datetime import time
        
        schedule, created = AreaSchedule.objects.get_or_create(
            area=area,
            defaults={
                'monday_start': time(8, 0),
                'monday_end': time(17, 0),
                'monday_active': True,
                'tuesday_start': time(8, 0),
                'tuesday_end': time(17, 0),
                'tuesday_active': True,
                'wednesday_start': time(8, 0),
                'wednesday_end': time(17, 0),
                'wednesday_active': True,
                'thursday_start': time(8, 0),
                'thursday_end': time(17, 0),
                'thursday_active': True,
                'friday_start': time(8, 0),
                'friday_end': time(17, 0),
                'friday_active': True,
                'saturday_active': False,
                'sunday_active': False,
                'grace_period_minutes': 15
            }
        )
        return schedule
    
    @staticmethod
    def is_work_day(area, date):
        """
        Verifica si una fecha es un día laboral para un área
        
        Args:
            area: Instancia del modelo Area
            date: Fecha a verificar
            
        Returns:
            bool: True si es día laboral, False en caso contrario
        """
        if not hasattr(area, 'schedule'):
            return False
        
        schedule = area.schedule
        weekday = date.weekday()
        
        if weekday == 0:
            return schedule.monday_active
        elif weekday == 1:
            return schedule.tuesday_active
        elif weekday == 2:
            return schedule.wednesday_active
        elif weekday == 3:
            return schedule.thursday_active
        elif weekday == 4:
            return schedule.friday_active
        elif weekday == 5:
            return schedule.saturday_active
        elif weekday == 6:
            return schedule.sunday_active
        
        return False
    
    @staticmethod
    def get_grace_period(area):
        """
        Obtiene el período de gracia para un área
        
        Args:
            area: Instancia del modelo Area
            
        Returns:
            int: Minutos de tolerancia
        """
        if hasattr(area, 'schedule'):
            return area.schedule.grace_period_minutes
        return 15  # Valor por defecto

File: admin-frontend/src/services/scheduleService.js
import api from './api'

class ScheduleService {
  /**
   * Obtener horario de un área específica
   */
  async getSchedule(areaId) {
    try {
      const response = await api.get(`/area-schedules/?area_id=${areaId}`)
      return response.data
    } catch (error) {
      console.error('Error obteniendo horario:', error)
      throw error
    }
  }

  /**
   * Crear horario para un área
   */
  async createSchedule(scheduleData) {
    try {
      const response = await api.post('/area-schedules/', scheduleData)
      return response.data
    } catch (error) {
      console.error('Error creando horario:', error)
      throw error
    }
  }

  /**
   * Actualizar horario existente
   */
  async updateSchedule(scheduleId, scheduleData) {
    try {
      const response = await api.patch(`/area-schedules/${scheduleId}/`, scheduleData)
      return response.data
    } catch (error) {
      console.error('Error actualizando horario:', error)
      throw error
    }
  }

  /**
   * Crear horario por defecto para un área
   */
  async createDefaultSchedule(areaId) {
    try {
      const response = await api.post(`/area-schedules/${areaId}/create_default/`)
      return response.data
    } catch (error) {
      console.error('Error creando horario por defecto:', error)
      throw error
    }
  }

  /**
   * Eliminar horario de un área
   */
  async deleteSchedule(scheduleId) {
    try {
      const response = await api.delete(`/area-schedules/${scheduleId}/`)
      return response.data
    } catch (error) {
      console.error('Error eliminando horario:', error)
      throw error
    }
  }

  /**
   * Validar horario antes de enviar
   */
  validateSchedule(schedule) {
    const errors = []
    
    // Validar que al menos un día esté activo
    const activeDays = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday']
    const hasActiveDay = activeDays.some(day => schedule[`${day}_active`])
    
    if (!hasActiveDay) {
      errors.push('Al menos un día debe estar activo')
    }

    // Validar horarios de días activos
    activeDays.forEach(day => {
      if (schedule[`${day}_active`]) {
        const start = schedule[`${day}_start`]
        const end = schedule[`${day}_end`]
        
        if (!start || !end) {
          errors.push(`${day} debe tener hora de inicio y fin`)
        } else if (start >= end) {
          errors.push(`${day} la hora de inicio debe ser menor que la de fin`)
        }
      }
    })

    // Validar tolerancia
    if (schedule.grace_period_minutes < 0 || schedule.grace_period_minutes > 120) {
      errors.push('La tolerancia debe estar entre 0 y 120 minutos')
    }

    return errors
  }

  /**
   * Crear horario por defecto (8:00 AM - 5:00 PM, L-V)
   */
  createDefaultScheduleData() {
    return {
      monday_start: '08:00',
      monday_end: '17:00',
      monday_active: true,
      tuesday_start: '08:00',
      tuesday_end: '17:00',
      tuesday_active: true,
      wednesday_start: '08:00',
      wednesday_end: '17:00',
      wednesday_active: true,
      thursday_start: '08:00',
      thursday_end: '17:00',
      thursday_active: true,
      friday_start: '08:00',
      friday_end: '17:00',
      friday_active: true,
      saturday_start: null,
      saturday_end: null,
      saturday_active: false,
      sunday_start: null,
      sunday_end: null,
      sunday_active: false,
      grace_period_minutes: 15
    }
  }
}

export default new ScheduleService()
